cpp/include/raft/core/device_resources_manager.hpp

```diff
@@ -0,0 +1,580 @@
+  {
+    auto lock = get_lock();
+    if (params_finalized_) {
+      RAFT_LOG_WARN(
+        "Attempted to set device_resources_manager properties after resources have already been "
+        "retrieved");
+    } else {
+      params_.workspace_allocation_limit.emplace(memory_limit);
+    }
+  }
+
+  // Thread-safe setter for the maximum memory pool size
+  void set_max_mem_pool_size_(std::optional<std::size_t> memory_limit)
+  {
+    auto lock = get_lock();
+    if (params_finalized_) {
+      RAFT_LOG_WARN(
+        "Attempted to set device_resources_manager properties after resources have already been "
+        "retrieved");
+    } else {
+      if (memory_limit) {
+        params_.max_mem_pool_size.emplace(*memory_limit);
+      } else {
+        params_.max_mem_pool_size = thrust::nullopt;
+      }
+    }
+  }
+
+  // Thread-safe setter for the initial memory pool size
+  void set_init_mem_pool_size_(std::optional<std::size_t> init_memory)
+  {
+    auto lock = get_lock();
+    if (params_finalized_) {
+      RAFT_LOG_WARN(
+        "Attempted to set device_resources_manager properties after resources have already been "
+        "retrieved");
+    } else {
+      if (init_memory) {
+        params_.init_mem_pool_size.emplace(*init_memory);
+      } else {
+        params_.init_mem_pool_size = thrust::nullopt;
+      }
+    }
+  }
+
+  // Retrieve the instance of this singleton
+  static auto& get_manager()
+  {
+    static auto manager = device_resources_manager{};
+    return manager;
+  }
+
+ public:
+  /**
+   * @brief Retrieve device_resources to be used with the RAFT API
+   *
+   * This thread-safe method ensures that a `device_resources` object with
+   * the same underlying stream and stream pool is returned every time it is
+   * called by the same host thread. This means that if `get_device_resources` is
+   * used to provide all `device_resources` in an application, then
+   * `raft::get_device_resources().sync_stream()` and (if a stream pool is used)
+   * raft::get_device_resources().sync_stream_pool() are guaranteed to synchronize all
+   * work previously submitted to the device by this host thread.
+   *
+   * If the max memory pool size set with `set_max_mem_pool_size` is non-zero,
+   * the first call of this method will also create a memory pool to be used
+   * for all RMM-based allocations on device.
+   *
+   * @param int device_id If provided, the device for which resources should
+   * be returned. Defaults to active CUDA device.
+   * @param workspace_mr If provided, a separate memory resource to be used
+   * for allocating temporary workspaces in RAFT calls.
+   */
+  static auto get_device_resources(int device_id = device_setter::get_current_device(),
+                                   std::shared_ptr<rmm::mr::device_memory_resource> workspace_mr = {
+                                     nullptr})
+  {
+    auto const& components = get_manager().get_device_components(device_id);
```

Comment:
From a larger overarching design standpoint, I'd have expected to see this manager actually manage a series of device resources objects, which might each have their own varying state, rather than as a "super factory" that just produces new instances. For one- this is assuming that the creation of the instances themselves is cheap enough that it's okay to do this often. Another problem with this is that it sidesteps the individual device resources from being able to maintain their own independent state. I don't mind that the `resources_manager` is the curator for consolidating the initial settings, but I'd prefer to have the resources manager actually manage a collection of `device_resources` instances, rather than just producing new instances each time they are needed. 

For example, FAISS does something very similar to what I'm suggesting, and I think it better defines the relationship and separation of concerns between the manager of instances and the actual instances.

Reply:
If we want to cache the `device_resources`, that's fine. I had considered doing so but did not like that it automatically extended the lifetime of the resources object beyond what might actually be used by the caller. I figured that we would prefer this to _just_ generate the `device_resources` object and then have the consumer use that object as it is normally used by consumers of the RAFT API. It feels like caching is adding an additional complication to do something that is already handled by the `device_resources` object. I'll add it for now and we can revert if we don't like it.